### PR TITLE
Fix and add test for empty chain and reorg consistency for gettxoutsetinfo.

### PR DESCRIPTION
**Pull request by gmaxwell**

### CODE REVIEW

**Comment by gmaxwell** [comment]
This will crash on master. Fix is currently in the per-txout PR.

**Comment by sipa** [comment]
utACK 4b3dd975a303a0d7efaec790c387f0285a4bb8c4

**Review by jonasschnelli — APPROVED** [review]
utACK 4b3dd975a303a0d7efaec790c387f0285a4bb8c4

**Comment by paveljanik** [comment]
ACK https://github.com/bitcoin/bitcoin/commit/4b3dd975a303a0d7efaec790c387f0285a4bb8c4

FWIW: The new test crashes on master with:

```
2017-05-24 06:50:24.034000 TestFramework (INFO): Initializing test directory /var/folders/65/fn0h49r55k7779vg1b_h461r0000gn/T/test0wr7ofxj
Assertion failed: (valid_), function key, file leveldb/db/db_iter.cc, line 67.
```

Fixed in this PR.

File: test/functional/blockchain.py [hunk 1]
```diff
@@ -53,6 +53,25 @@ def _test_gettxoutsetinfo(self):
         assert_equal(len(res['bestblock']), 64)
         assert_equal(len(res['hash_serialized']), 64)
 
+        b1hash = node.getblockhash(1)
```
Comment (jnewbery) [inline test/functional/blockchain.py]: Can you add a comment (or info log) here explaining what this test is for:

```py
self.log.info("Test that gettxoutsetinfo() works for blockchain with just the genesis block")
```

File: test/functional/blockchain.py [hunk 2]
```diff
@@ -53,6 +53,25 @@ def _test_gettxoutsetinfo(self):
...
+        node.invalidateblock(b1hash)
+
+        res2 = node.gettxoutsetinfo()
+        assert_equal(res2['transactions'], 0)
+        assert_equal(res2['total_amount'], Decimal('0'))
+        assert_equal(res2['height'], 0)
+        assert_equal(res2['txouts'], 0)
```
Comment (jnewbery) [inline test/functional/blockchain.py]: Why not test all fields:

```py
assert_equal(res2['bestblock'], node.getblockhash(0)))
assert_equal(len(res2['hash_serialized']), 64)
```

File: test/functional/blockchain.py [hunk 3]
```diff
@@ -53,6 +53,25 @@ def _test_gettxoutsetinfo(self):
...
+
+        node.reconsiderblock(b1hash)
+
```
Comment (jnewbery) [inline test/functional/blockchain.py]: Again, a comment/info log would be nice here:

```py
self.log.info("Test that gettxoutsetinfo() returns the same result after invalidate/reconsider block")
```

File: test/functional/blockchain.py [hunk 4]
```diff
@@ -53,6 +53,25 @@ def _test_gettxoutsetinfo(self):
...
+        res3 = node.gettxoutsetinfo()
+        assert_equal(res['total_amount'], res3['total_amount'])
```
Comment (jnewbery) [inline test/functional/blockchain.py]: Consider just testing equality of res and res3 for succinctness:

```py
assert_equal(res, res3)
```

**Review by jnewbery — COMMENTED** [review]
I can't review the code change in src/txdb.cpp since I don't know that code, but I can confirm that this test causes bitcoind to crash without the code change, and passes successfully with the code change.

A few nits inline. Also consider changing line 35 of the test from:

```py
        self.num_nodes = 2
```
to:

```py
        self.num_nodes = 1
```

I know that wsan't changed by this PR, but 2 nodes are unnecessary for this test, and changing it to 1 cuts the test time in half (from 5s to 2.5s on my pc) due to node startup/shutdown time.

**Comment by sdaftuar** [comment]
ACK

File: test/functional/blockchain.py (hunk 4 above)
Comment (gmaxwell) [inline test/functional/blockchain.py]: Equality will be the wrong comparison with the disk usage field which is about to be added.

File: test/functional/blockchain.py (hunk 1 above)
Comment (gmaxwell) [inline test/functional/blockchain.py]: done

File: test/functional/blockchain.py (hunk 2 above)
Comment (gmaxwell) [inline test/functional/blockchain.py]: done (and also added that to the first check)

File: test/functional/blockchain.py (hunk 3 above)
Comment (gmaxwell) [inline test/functional/blockchain.py]: done

**Comment by gmaxwell** [comment]
updated for @jnewbery 's nits.  (Thanks!)

**Comment by jnewbery** [comment]
tested reACK https://github.com/bitcoin/bitcoin/pull/10445/commits/513da90cdd8463964e8c430b25f4326defc0bc96

**Comment by sdaftuar** [comment]
utACK 513da90

**Comment by paveljanik** [comment]
reACK 513da90

**Pull request merged**